cpp/node_similarity_module/algorithms/node_similarity.hpp

```diff
@@ -0,0 +1,193 @@
+#include <mgp.hpp>
+#include <math.h>
+
+
+namespace node_similarity_util {
+
+struct pair_hash {
+    template <class T1, class T2>
+    std::size_t operator () (std::pair<T1, T2> const &pair) const
+    {
+        std::size_t h1 = std::hash<T1>()(pair.first);
+        std::size_t h2 = std::hash<T2>()(pair.second);
+
+        return h1 ^ h2;
+    }
+};
+
+enum Similarity { jaccard, overlap, cosine };
+
+
+}  // namespace node_similarity_util
+
+
+namespace node_similarity_algs {
+
+
+// Lambda functions for computing concrete similarities
+/*
+Calculates Jaccard similarity between two nodes based on their neighbours.
+ns1 are neighbours of the first node
+ns2 are neighbours of the second node
+*/
+
+double JaccardFunc(const std::set<uint64_t> &ns1, const std::set<uint64_t> &ns2) {
+    std::set<uint64_t> elem_union, elem_intersection;
+    std::set_intersection(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_intersection, elem_intersection.begin()));
+    std::set_union(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_union, elem_union.begin()));
+    if (elem_union.size() == 0) {
+        return 0.0;
+    }
+    return elem_intersection.size() / (double) elem_union.size();
+}
+
+/*
+Calculates overlap function between two set of neighbours.
+ns1 are neighbours of the first node.
+ns2 are neighbours of the second node.
+*/
+double OverlapFunc(const std::set<uint64_t> &ns1, const std::set<uint64_t> &ns2) {
+    std::set<uint64_t> elem_intersection;
+    std::set_intersection(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_intersection, elem_intersection.begin()));
+    int denominator = std::min(ns1.size(), ns2.size());
+    if (denominator == 0) {
+        return 0.0;
+    }
+    return elem_intersection.size() / (double) denominator;
+}
+
+
+double CosineFunc(const mgp::List &prop1, const mgp::List &prop2) {
+    int size = prop1.Size();
+    double similarity = 0.0, node1_sum = 0.0, node2_sum = 0.0;
+    for (int i = 0; i < size; ++i) {
+        double val1 = prop1[i].ValueDouble(), val2 = prop2[i].ValueDouble();
+        similarity += val1 * val2;
+        node1_sum += val1 * val1;
+        node2_sum += val2 * val2;
+    }
+    double denominator = sqrt(node1_sum) * sqrt(node2_sum);
+    if (denominator < 1e-9) {
+        return 0.0;
+    } 
+    return similarity / denominator;
+
+}
+
+
+
+/*
+Calculates cosine similarity function between two nodes for a given property.
+*/
+double CosineFuncWrapper(const mgp::Node &node1, const mgp::Node &node2, const std::string &property) {
+    const auto &prop1_it = node1.GetProperty(property);
+    const auto &prop2_it = node2.GetProperty(property);
+    if (prop1_it.IsNull() || prop2_it.IsNull()) {
+        throw mgp::ValueException("All nodes should have property " + property);
```

Comment:
```suggestion
        throw mgp::ValueException(f"All nodes should have the {property} property");
```

Reply:
this is not python :joy: I don't think this works in cpp